File: src/tests/test_page.rb
# frozen_string_literal: true

require_relative '../lib'

class TestPage < Test::Unit::TestCase
  def test_filename
    page = Blog::Page.new('something.md')
    assert_equal(page.filename, 'something.html')
    page = Blog::Page.new('something.html')
    assert_equal(page.filename, 'something.html')
    page = Blog::Page.new('entry/thing/something.md')
    assert_equal(page.filename, 'something.html')
  end

  def test_pathname
    page = Blog::Page.new('something.md')
    assert_equal(page.pathname, '/something.html')
    page = Blog::Page.new('something.html')
    assert_equal(page.pathname, '/something.html')
    page = Blog::Page.new('entry/thing/something.md')
    assert_equal(page.pathname, '/something.html')
  end

  def test_frontmatter
    content = <<~CONTENT.strip
      ---
      title: something
      nav: 1
      ---

      This is a test entry!
    CONTENT
    page = Blog::Page.new('test.md', content: content)
    assert_equal(
      page.frontmatter,
      {
        'title' => 'something',
        'nav' => 1
      }
    )
  end

  def test_nav?
    content = <<~CONTENT.strip
      ---
      title: something
      nav: 1
      ---

      This is a test entry!
    CONTENT
    page = Blog::Page.new('test.md', content: content)
    assert_true(page.nav?)
  end

  def test_entry?
    page = Blog::Page.new('_posts/entry.md')
    assert_true(page.entry?)
    page = Blog::Page.new('_pages/page.md')
    assert_false(page.entry?)
  end

  def test_date
    assert_nil(Blog::Page.new('_pages/page.md').date)

    date = Blog::Page.new('_posts/2020-01-01-entry.md').date
    assert_equal(date, Date.new(2020, 1, 1))
  end

  def test_title
    page = Blog::Page.new(
      '_pages/page.md',
      content: <<~FILE
        ---
        title: A Test Title
        ---
      FILE
    )
    assert_equal(page.title, 'A Test Title')
    entry = Blog::Page.new('_posts/2020-01-01-entry.md')
    assert_equal(entry.title, 'Wednesday, January 1 2020')
  end

  def test_description
    page = Blog::Page.new(
      '_pages/page.md',
      content: <<~FILE
        ---
        description: a test description
        ---
      FILE
    )
    assert_equal(page.description, 'a test description')
    entry = Blog::Page.new(
      '_posts/2020-01-01-entry.md',
      content: <<~FILE
        ---
        title: a fart in the wind
        ---
      FILE
    )
    assert_equal(entry.description, 'a fart in the wind')
  end
end

File: src/models/test/test_page.py
import tempfile
import unittest

from ..page import Page


class TestPage(unittest.TestCase):
    def test_filename(self):
        actual = Page('./pages/index.html').filename
        expected = 'index.html'
        self.assertEqual(actual, expected)

        actual = Page('./pages/index.html.j2').filename
        expected = 'index.html'
        self.assertEqual(actual, expected)

    def test_is_entry(self):
        self.assertFalse(Page('./pages/index.html').is_entry)
        self.assertFalse(Page('./pages/index.html.j2').is_entry)
        self.assertFalse(Page('./pages/index.html.j2').is_entry)
        self.assertTrue(Page('./entries/something.html').is_entry)

    def test_date(self):
        actual = Page('./entries/2023-08-10.html').date
        actual = (actual.year, actual.month, actual.day)
        expected = (2023, 8, 10)
        self.assertEqual(actual, expected)

    def test_metadata(self):
        content = '''
<!-- meta:title A Test Title -->
<!-- meta:description A Test Description -->
        '''

        tmp = tempfile.NamedTemporaryFile()

        with open(tmp.name, 'w') as f:
            f.write(content)

        actual = Page(tmp.name).metadata
        expected = {
            'title': 'A Test Title',
            'description': 'A Test Description',
        }
        self.assertEqual(actual, expected)
